### ISSUE TITLE
Paper in details

### ISSUE THREAD

**Issue by JustinKai0527** [yuhaoliu7456/Diff-Plugin#6]
Very interesting work!!!
Q1.  I want to ask the section 3.3. Task-Plugin, you want to get task-specific visual guidance prior and spatial prior to learn the part we want to remove and preserve the image, but the training loss initial latent is encoded by ground truch image which is "clean" image? How to let the TPB&SCB learn to clean the unclean image?
Q2.  I want to check that my understanding about the overall pipeline. So I give a image and text that I want to do desnow, the image will feed to task-plugin to get two prior, and then using plugin-selector to compute similarity between the text. And choose the prior which similarity exceeds the threshold, feed the task-specific visual guidance prior to cross-attn, spatial prior to final stage of decoder?
Thx!!!

**Comment by yuhaoliu7456** [comment]
Thanks for your interest. 

For Q1, during the diffusion training, we regard the GT as $x_{0}$ and add noise at different extents. As the input of our Task-Plugin is a degraded image, the whole diffusion training is actually a conditional removal process, which has to learn how to model the relationship between the input degraded image and noisy GT.

For Q2, yes, your understanding is exactly right.

**Comment by JustinKai0527** [comment]
Thanks for quickly reponding!
In Q1, what I got confused about is that the GT image, I think, is a clean image, which means it is not a snow or rain image. However, you can train the task-plugin to learn how to clean up the unwanted details. Why? 
From my understanding, the loss term is basically a denoising term in diffusion model which learn how to reconstruct the image.

**Comment by yuhaoliu7456** [comment]
Because the input of the task-plugin is a degraded image(e.g., snow image). During the diffusion/reconstruction process, the snow image is actually a conditional input, and this force the model to get rid of the snow pattern from the task-plugin such that the reconstruction process can be done.
________________________________
发件人: Justin Kai ***@***.***>
发送时间: Monday, April 29, 2024 12:12:54 PM
收件人: yuhaoliu7456/Diff-Plugin ***@***.***>
抄送: YuhaoLiu ***@***.***>; Mention ***@***.***>
主题: Re: [yuhaoliu7456/Diff-Plugin] Paper in details (Issue #6)


@yuhaoliu7456<https://github.com/yuhaoliu7456> Thanks for quickly reponding!
In Q1, what I got confused about is that the GT image, I think, is a clean image, which means it is not a snow or rain image. However, you can train the task-plugin to learn how to clean up the unwanted details. Why?
From my understanding, the loss term is basically a denoising term in diffusion model which learn how to reconstruct the image.

―

**Comment by JustinKai0527** [comment]
Okay, so you have a paired image that is snow image and w/o snow image.
![image](https://github.com/yuhaoliu7456/Diff-Plugin/assets/103019623/e91eed5a-235e-483a-983e-3bb2107b2f17)
In the picture, you feed snow image to get the prior, and the I_hat in loss term below is w/o snow image?
![image](https://github.com/yuhaoliu7456/Diff-Plugin/assets/103019623/de40b4a1-848d-41e1-ab0b-6bd95f3b6c1d)
So in training, the prior got is snow image prior and try to recontruct the z<sub>t</sub> which is encoded by I_hat(w/o snow image)?

**Comment by lyf1212** [comment]
> @yuhaoliu7456 Thanks for replying. Okay, so you have a paired image that is snow image and w/o snow image. ![image](https://private-user-images.githubusercontent.com/103019623/326328986-e91eed5a-235e-483a-983e-3bb2107b2f17.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTYwMzU5NDQsIm5iZiI6MTcxNjAzNTY0NCwicGF0aCI6Ii8xMDMwMTk2MjMvMzI2MzI4OTg2LWU5MWVlZDVhLTIzNWUtNDgzYS05ODNlLTNiYjIxMDdiMmYxNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUxOFQxMjM0MDRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05Y2M5NjA3YTJjMTlmYTMzMDEyZDYwODIyNzNkNjMwMGZmOWM3MjlhMWY4YTMzOGI3NDIxMjNmNTNlYTdmYTY0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.dWojIPKcdfWYAMed7HTDQRzRjKO-B73aEHHOG_Y89j4) In the picture, you feed snow image to get the prior, and the I_hat in loss term below is w/o snow image? ![image](https://private-user-images.githubusercontent.com/103019623/326329089-de40b4a1-848d-41e1-ab0b-6bd95f3b6c1d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTYwMzU5NDQsIm5iZiI6MTcxNjAzNTY0NCwicGF0aCI6Ii8xMDMwMTk2MjMvMzI2MzI5MDg5LWRlNDBiNGExLTg0OGQtNDFlMS1hYjBiLTZiZDk1ZjNiNmMxZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUxOFQxMjM0MDRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mMDcyZjQ3MDNiMWIzYzE4OTRlZjgyMjRjM2JhOThhY2ZmODBmZWViMDgyODRmZjQ4YWQyNzZiZWM3ZTQ0ZjNkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.-DpjiWvL1tN7khvUdPCLST96QGCFrL9wcdPxZbofAc0) So in training, the prior got is snow image prior and try to recontruct the zt which is encoded by I_hat(w/o snow image)?

I think it is the right understanding. In other words, you can also treat this method as a unified ControlNet, which basically takes the snowy/rainy/noisy... images as conditional input of ControlNet and guide the diffusion process by the middle features insertion and the genious design of zero-conv by ControlNet architechture. The main contribution of this work is they used CLIP Image Encoder to extract the visual embedding of those degradation images which will be aligned with those "degradation descriptions" and designed a task selector by a basic contrastive loss and a newly proposed dataset. Also looking forward for the official reply from the author~

**Comment by yuhaoliu7456** [comment]
The main contribution of our work is the whole framework, not just the control-based encoder. Please take a look at our main paper for more detailed idea. If there are no other questions, I will close this comment later. Thanks.

**Comment by How-Wang** [comment]
> @yuhaoliu7456 Thanks for quickly responding! In Q1, what I got confused about is that the GT image, I think, is a clean image, which means it is not a snow or rain image. However, you can train the task-plugin to learn how to clean up the unwanted details. Why? From my understanding, the loss term is basically a denoising term in diffusion model which learn how to reconstruct the image.

I think your understanding and concern are correct and the "control-based encoder" design is extremely strange. I think they just don't know how to explain the design of "use a degradation feature to clean a image without degradation", so they say "Oh! our contribution is the whole framework! To see our paper for detailed again!" Hahaha

**Comment by How-Wang** [comment]
> > @yuhaoliu7456 Thanks for replying. Okay, so you have a paired image that is snow image and w/o snow image. ![image](https://private-user-images.githubusercontent.com/103019623/326328986-e91eed5a-235e-483a-983e-3bb2107b2f17.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTYwMzU5NDQsIm5iZiI6MTcxNjAzNTY0NCwicGF0aCI6Ii8xMDMwMTk2MjMvMzI2MzI4OTg2LWU5MWVlZDVhLTIzNWUtNDgzYS05ODNlLTNiYjIxMDdiMmYxNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUxOFQxMjM0MDRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05Y2M5NjA3YTJjMTlmYTMzMDEyZDYwODIyNzNkNjMwMGZmOWM3MjlhMWY4YTMzOGI3NDIxMjNmNTNlYTdmYTY0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.dWojIPKcdfWYAMed7HTDQRzRjKO-B73aEHHOG_Y89j4) In the picture, you feed snow image to get the prior, and the I_hat in loss term below is w/o snow image? ![image](https://private-user-images.githubusercontent.com/103019623/326329089-de40b4a1-848d-41e1-ab0b-6bd95f3b6c1d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTYwMzU5NDQsIm5iZiI6MTcxNjAzNTY0NCwicGF0aCI6Ii8xMDMwMTk2MjMvMzI2MzI5MDg5LWRlNDBiNGExLTg0OGQtNDFlMS1hYjBiLTZiZDk1ZjNiNmMxZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUxOFQxMjM0MDRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mMDcyZjQ3MDNiMWIzYzE4OTRlZjgyMjRjM2JhOThhY2ZmODBmZWViMDgyODRmZjQ4YWQyNzZiZWM3ZTQ0ZjNkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.-DpjiWvL1tN7khvUdPCLST96QGCFrL9wcdPxZbofAc0) So in training, the prior got is snow image prior and try to recontruct the zt which is encoded by I_hat(w/o snow image)?
> 
> I think it is the right understanding. In other words, you can also treat this method as a unified ControlNet, which basically takes the snowy/rainy/noisy... images as conditional input of ControlNet and guide the diffusion process by the middle features insertion and the genious design of zero-conv by ControlNet architechture. The main contribution of this work is they used CLIP Image Encoder to extract the visual embedding of those degradation images which will be aligned with those "degradation descriptions" and designed a task selector by a basic contrastive loss and a newly proposed dataset. Also looking forward for the official reply from the author~

I can understand your explanation. However, The controlNet with SD only contain one task such as scribble to image. Why the model can specify each Plugin's TPB belonging to correct degradation type through training? The loss function design doesn't contain the degradation image, but a clean image w/o degradation (with noise). Do you have any idea?

**Comment by How-Wang** [comment]
Initially, 𝑧0 does not have degradation (though it contains noise), and the final clean image 𝑧𝑡 also does not have degradation. Why they start with the GT image? I think the key point during inference is that this paper does not employ DDIM inversion!

Therefore, TPB is not actually about "removing" degradation but guiding the model to avoid generating degradation. The information from SCB which containing actually spatial info and hoping remove the degradation info by SCB itself is feed in the last step.

By using 𝐹𝑝 which initially contains both image content and degradation information, we train to isolate only the degradation information. This allows the denoising process to understand what the degradation looks like and avoid generating it.
